promql: pass down subquery interval

GiedriusS opened the pull request:
If we are populating series for a subquery then set the interval parameter accordingly so that downstream users could use that
information.

Fixes #11162. Tested on Thanos Query with this forked version that pushed down subqueries are properly evaluated.

Changes seem obvious from Prometheus perspective so I'm not sure if any tests are needed. I think e2e tests will be needed for this on Thanos side.

Signed-off-by: Giedrius Statkevičius <giedrius.statkevicius@vinted.com>

codesome commented on promql/engine.go:
```diff
@@ -823,6 +823,17 @@ func (ng *Engine) getTimeRangesForSelector(s *parser.EvalStmt, n *parser.VectorS
 	return start, end
 }
 
+func (ng *Engine) getSubqueryInterval(path []parser.Node) time.Duration {
```
```suggestion
func (ng *Engine) getLastSubqueryInterval(path []parser.Node) time.Duration {
```

codesome replied:
Just to confirm: you only need the inner-most interval because that is technically the required step for the series. (I see that is what is being done)

GiedriusS replied:
Yep, that's true. I noticed that after the tests started failing and only then did I think about this case.

codesome commented on promql/engine.go:
```diff
@@ -833,10 +844,14 @@ func (ng *Engine) populateSeries(querier storage.Querier, s *parser.EvalStmt) {
 		case *parser.VectorSelector:
 			start, end := ng.getTimeRangesForSelector(s, n, path, evalRange)
+			interval := s.Interval
+			if interval == 0 {
```
Should you not always try to get the subquery interval and set that if it is non-zero? Because that is what will be applied to the vector selector in that case. Can you try a query on Thanos with query interval and subquery interval are incompatible (i.e. not divisible by each other).

codesome replied:
s.Interval will be 0 only in the case of an instant query I believe

codesome replied:
Even in case of a range query, the samples required for a subquery will be aligned with the subquery interval and not the range query interval (think of subquery as yet another range query with some timestamp alignments). So I meant literally only taking innermost subquery interval iff it exists. Was there any specific reason to prefer the range query interval if there was a subquery?

GiedriusS replied:
I did not know that. Let me update my branch.

GiedriusS commented:
Added repro on Thanos side https://github.com/thanos-io/thanos/pull/5602 + a temporary fork to show that the tests fail without this change. While working on tests, I noticed that subqueries were not handled which don't have a `step` parameter. I have now fixed this.

GiedriusS commented:
Ping :hugs:

GiedriusS commented:
Fixed according to suggestions @codesome :hugs:

codesome reviewed (approved):
LGTM

wanghaao commented:
> bqueries

I want to know why not create a new field ``SubQueryStep``. It not only passes down the interval of the SubQuery, but differs subquery from other scenes.

codesome commented:
> I want to know why not create a new field SubQueryStep. It not only passes down the interval of the SubQuery, but differs subquery from other scenes.

Because for the use case we only needed to know what is the interval at which the series was required. Subquery or not doesn't matter.
